21 metainformation in generated classes

Wauplin opened the pull request:
Following the discussion started in #21 .

- we only use information included in the JSON (e.g : the urls)
- meta data is stored in a `FHIRMeta` NamedTuple object. @julian-r mentioned the idea of using the already existing `Meta` resource implemented by FHIR. The problem is that `Meta` depends on `Element` which depends on `Extension` which depends on a lot of FHIR objects (`Coding`, `Range`, `Address`, `Age`, `Annotation`,....). And since I need to instanciate the `Meta` resource when I declare each classes, I got some validation issues (I can't do the `update_forward_refs()` call on `Meta` while all the other classes are not defined yet. A possible solution would be to declare all of the classes first, do the `update_forward_refs()` calls and then dynamically add the meta information to the classes in the very end of the `r4.py` file. I felt that this solution was not very pythonic/readable since metainformation would not be declared at the same time as the classes. Please let me know if the NamedTuple solution is ok for you or if you prefer a proper FHIR resource `Meta`.
- In case of `DocEnum`, I have added the URL to the docstring but that's all. I will create a new issue about this specific topic.
- (this PR also removes the commented properties in `resource.py.jinja2 `. Hope it wasn't mandatory)

Wauplin committed:
Add profile urls property to FHIRClass

Wauplin committed:
Add codesystem url in generated docstring

Wauplin committed:
Add FHIRMeta object + populate profile urls in generated resources

Wauplin commented:
Example of generated code in `r4.py` : 

**EDIT (13th July) :** Updated according to last commit.

```py
(...) 

class FHIRAbstractBase(pydantic.BaseModel):
    """Abstract base class for all FHIR elements.
    """
    class META:
        profile: typing.List[str] = []
        """ Profiles this resource claims to conform to.
        List of `str` items. """

    def dict(self, *args, **kwargs):
        serialized = super().dict(*args, **kwargs)
        return _without_empty_items(serialized)

    class Config:
        alias_generator = camelcase_alias_generator
        allow_population_by_field_name = True

(...) 

class AbstractType(str, DocEnum):
    """ Defining URL : http://hl7.org/fhir/abstract-types
    """
    type = "Type", """A place holder that means any kind of data type """
    any = "Any", """A place holder that means any kind of resource """

(...)

class MedicinalProductPharmaceutical(DomainResource):
    """ A pharmaceutical product described in terms of its composition and dose
    form.
    """
    resource_type = "MedicinalProductPharmaceutical"
    class META:
        profile: typing.List[str] =[
           "http://hl7.org/fhir/StructureDefinition/MedicinalProductPharmaceutical",
        ]
        """ Profiles this resource claims to conform to.
        List of `str` items. """

    identifier: typing.Optional[typing.List["Identifier"]]
    """ An identifier for the pharmaceutical medicinal product.
    List of `Identifier` items (represented as `dict` in JSON). """

    (...)

(...)
```

julian-r commented:
Well yes the information is now there, my suggestion was to go for something like this:

https://github.com/samuelcolvin/pydantic/issues/184#issuecomment-447638017

Because the access via `_META` looks like you are accessing a private member.

Wauplin committed:
Add profile as META class in generated resources

Wauplin committed:
Merge branch 'master' into 21-metainformation-in-generated-classes

Wauplin commented:
@julian-r : Made requested change to put Meta information as a class instead of a private attribute. I added a test to check that every generated class has indeed at least a related profile.

+ I changed my previous comment in this PR with the updated example.

Wauplin committed:
Merge branch 'master' into 21-metainformation-in-generated-classes
